### PR TITLE
[8.x] Add setRelation to Eloquent's collection

### PR DESCRIPTION
**Pull request by ahmedash95**
This PR adds a shortcut to set model relation from Eloquent Collection.

```php
# instead of 
$comments = $post->comments;
$comments->each->setRelation('post', $post);

# would be
$comments = $post->comments;
$comments->setRelation('post',$post);
```

also setting many relations would be easy and nice 

```php
# instead of 
$products->setRelation('category', $category);
$products->setRelation('brand', $brand)

# would be
$products->setRelation([ 
  'category' => $category
  'brand' => $brand
]);
```

### CODE REVIEW

File: src/Illuminate/Database/Eloquent/Collection.php [hunk 1]
```diff
@@ -237,6 +237,24 @@ public function contains($key, $operator = null, $value = null)
         });
     }
 
+    /**
+     * Set realtion over each item.
+     *
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Collection.php:242]: Missing phpdoc.

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: src/Illuminate/Database/Eloquent/Collection.php [hunk 2]
```diff
@@ -240,6 +240,8 @@ public function contains($key, $operator = null, $value = null)
     /**
      * Set realtion over each item.
      *
+     * @param  mixed  $name
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Eloquent/Collection.php:243]: Why is this mixed?

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: src/Illuminate/Database/Eloquent/Collection.php (hunk 2 above)
Comment (ahmedash95) [reply src/Illuminate/Database/Eloquent/Collection.php:243]: to support an array of relations, check second test case 
```php
$model->setRelation([
   'a' => $relationA,
   'b' => $relationB,
])
```

Comment (GrahamCampbell) [reply src/Illuminate/Database/Eloquent/Collection.php:243]: so not mixed? `string|array`?

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: src/Illuminate/Database/Eloquent/Collection.php (hunk 2 above)
Comment (ahmedash95) [reply src/Illuminate/Database/Eloquent/Collection.php]: make sense

**Comment by morloderex** [comment]
Hm, have you considered the possiblity of a collection with different kind of models something like a morped collection?

**Comment by anouarabdsslm** [comment]
Nice. Wonder if is it better to handle the implementation with `setRelation` for single relation and `setRelations`  for bulk, so we follow the same philosophy as the model implementation.

**Comment by ahmedash95** [comment]
> Hm, have you considered the possiblity of a collection with different kind of models something like a morped collection?

I do not see any problem with that. Is there something I'm missing?

**Comment by ahmedash95** [comment]
> Nice. Wonder if is it better to handle the implementation with `setRelation` for single relation and `setRelations` for bulk, so we follow the same philosophy as the model implementation.

Actually that's a good point. but I prefer to have one shortcut for both. Let's wait for Taylor to reply and see if I've to change it. otherwise, we can submit another PR for model's `setRelation` to apply the same behavior

**Comment by taylorotwell** [comment]
I really think `each->setRelation` is fine.

**Pull request closed without merging**